File: level1/main.py

import json


def readData(pathFileInput):
    """Read a json file

    :param pathFileInput: the path of the input file
    :type pathFileInput: str
    :return: json array

    """
    with open(pathFileInput, 'r') as f:
        data = json.load(f)

    return data


def calculComissions(sumAmount, numberDeals):
    """calculation of commission on the total amount

    :param sumAmount: total sales amount  for each user
    :type sumAmount: float
    :param numberDeals: number of total deals for each user
    :type numberDeals: int
    :return: a number
    :rtype: float
    """
    comission = 0
    if numberDeals < 3:
        comission = 0.1*sumAmount
    else:
        comission = 0.2*sumAmount
    if sumAmount > 2000:
        comission += 500
    return comission


def applyCalcul(data):
    """calculation of total sales amount  and number of dealsfor each user to calculate the commission 
    :param data: json Array
    :return: the id of user and his comissions
    :rtype: list
    """
    dictJson = []

    for user in data["users"]:
        userId = user["id"]
        # If the user don't have any deal yet, comission=0
        outputUser = [deal for deal in data['deals'] if deal['user'] == userId]
        totalAmount = sum(x["amount"] for x in outputUser)
        numberDeals = len(outputUser)
        comission = calculComissions(totalAmount, numberDeals)
        dictJson.append({"user_id": userId, "comission": comission})
    return dictJson


def dataToJson(output, pathFileOutput):
    """ write the output in a json file
    :param output: list of json data
    :return: dict

    """
    outputJson = {"comissions": output}
    #write in file.json
    with open(pathFileOutput, 'w') as outfile:
        outfile.write(json.dumps(outputJson, indent=4))
    return outputJson


if __name__ == "__main__":
    inputPath = 'data/input.json'
    outputPath = "data/output.json"
    # load data
    print("***Load Data**")
    data = readData(inputPath)
    print(data)
    # Calculation of commissions
    print("***Calcul des comissions***")
    comissions = applyCalcul(data)
    # Write in json file
    print("***Json output""")
    print(dataToJson(comissions, outputPath))

File: level1/main.js

/**
 * 
 * @param {String} path 
 * @returns 
 */
function readJsonFile(path) {
    try {
        var json = require(path);
        return json;
    }
    catch (err) { throw err; }
}



/**
 * 
 * @param {Number} TotalAmount 
 * @param {Number} numberDeals 
 * @returns 
 */
function calculComission(TotalAmount, numberDeals) {
    var comission = 0
    if (numberDeals < 3) {
        comission = 0.1 * TotalAmount;
    }
    else {
        comission = 0.2 * TotalAmount;
    }
    if (TotalAmount > 2000) {
        comission += 500;
    }
    return comission;
}
/**
 * 
 * @param {Json} data 
 * @returns 
 */
function applyCalcul(data) {
    var comissions = [];

    for (key in data.users) {
        var comission = 0;
        var userID = data.users[key].id;
        var outputUser = data['deals'].filter(x => x.user == userID);
        if (outputUser.length > 0) {
            var TotalAmount = outputUser.map(x => x.amount).reduce((a, b) => a + b);//calculate the sum of amount for each user
            var numberDeals = outputUser.length;
            var comission = calculComission(TotalAmount, numberDeals);
        }
        comissions.push({ user_id: userID, commission: comission });


    }
    return comissions;

}
/**
 * 
 * @param {Array} output 
 * @param {String} pathFileOutput 
 */
function dataToJson(output, pathFileOutput) {
    outputJson = { "comissions": output }
    outputData = JSON.stringify(outputJson);

    console.log(outputData);
    var fs = require("fs");
    fs.writeFile(pathFileOutput, outputData, (err) => {
        if (err) {
            throw err;
        }
        console.log("JSON data is saved.");
    });



}
const pathFileInput = './data/input.json'
const pathFileOutput  = './data/output.json';       

var dataUser = readJsonFile(pathFileInput);
comissions = applyCalcul(dataUser);
dataToJson(comissions, pathFileOutput);